### PR TITLE
Adding optional SOCKS proxy support

### PR DESCRIPTION
**Pull request by justasmalinauskas**
Hello, as I use SOCKS proxies to get content instead HTTP ones, I have added support to it. 

I also would like to hear feedback about it :)

### CODE REVIEW

File: cykdl/__main__.py [hunk 1]
```diff
@@ -145,10 +145,25 @@ def main():
...
-            'https': args.proxy
-        })
+        if args.proxy.startswith("socks"):
+            try:
+                import socks
+                from sockshandler import SocksiPyHandler
+            except ImportError:
+                logger.error('Failed to load PySocks module, if you use pip package manager you can install it with command `pip install PySocks`')
+                exit = 1
+            from urllib.parse import urlparse
```
Comment (SeaHOH) [inline cykdl/__main__.py]: It is necessary to import from compact module.

File: cykdl/__main__.py [hunk 2]
```diff
@@ -145,10 +145,25 @@ def main():
...
+            parsed_socks_proxy = urlparse(args.proxy)
+            if  parsed_socks_proxy.scheme == "socks4":
+                sockstype = socks.SOCKS4
+            if  parsed_socks_proxy.scheme == "socks5":
+                sockstype = socks.SOCKS5
+            proxy_handler = SocksiPyHandler(sockstype,  parsed_socks_proxy.hostname, parsed_socks_proxy.port, True, parsed_socks_proxy.username, parsed_socks_proxy.password)
```
Comment (SeaHOH) [inline cykdl/__main__.py]: `rdns` is not always effective, need to resolve host if got a `0x5b` exception.

File: cykdl/__main__.py [hunk 3]
```diff
@@ -145,10 +145,25 @@ def main():
         proxy_handler = ProxyHandler()
         args.proxy = os.environ.get('HTTP_PROXY', 'none')
     else:
-        proxy_handler = ProxyHandler({
-            'http': args.proxy,
-            'https': args.proxy
-        })
+        if args.proxy.startswith("socks"):
```
Comment (SeaHOH) [inline cykdl/__main__.py]: Don't wrap, use the `elif` keyword.

File: cykdl/__main__.py [hunk 4]
```diff
@@ -144,26 +144,21 @@ def main():
...
+            if parsed_socks_proxy.scheme == "socks5":
                 sockstype = socks.SOCKS5
-            proxy_handler = SocksiPyHandler(sockstype,  parsed_socks_proxy.hostname, parsed_socks_proxy.port, True, parsed_socks_proxy.username, parsed_socks_proxy.password)
-        else:
-            proxy_handler = ProxyHandler({
-                'http': args.proxy,
-                'https': args.proxy
-            })
+            proxy_handler = SocksiPyHandler(sockstype, parsed_socks_proxy.hostname, parsed_socks_proxy.port, False, parsed_socks_proxy.username, parsed_socks_proxy.password)
+        except (socks.SOCKS5Error, socks.SOCKS5AuthError, socks.SOCKS4Error, socks.GeneralProxyError) as e:
```
Comment (SeaHOH) [inline cykdl/__main__.py]: No exception will be raised by defined a handler.

`rdns` only supported in SOCKS4A & SOCKS5, no effect with SOCKS4, it returned `0x5B` when `rdns` is set. Catch this error and re-connect after hostname is resolved, that must patch `SocksiPyConnection` & `SocksiPyConnectionS` if used them.

You may send issues or pull requests  to PySocks, or only fixed it in `ykdl.compact`.

File: ykdl/compact.py [hunk 5]
```diff
@@ -18,6 +18,8 @@
     compact_bytes = bytes
     from urllib.parse import unquote as compact_unquote
     from urllib.parse import quote
+    import socks
+    from sockshandler import SocksiPyHandler
```
Comment (SeaHOH) [inline ykdl/compact.py]: I am talking about urlparse, not socks, if you has no plan to fix `SocksiPyHandler` yourself.

BTW: same import should put outside of py2/3 judge.

**Comment by SeaHOH** [comment]
I have made a PR to PySocks.

**Comment by justasmalinauskas** [comment]
@SeaHOH I will wait till your patches in PySocks gonna be merged and after that I will finish it up, that's okay?

**Comment by SeaHOH** [comment]
Ok, I predict it will take some days to passed.
If the patch is not passed, we can include it in the repository directly.

**Comment by Anorov** [comment]
I'd recommend not using `sockshandler.py`; we're no longer maintaining it, or HTTP proxy support in general, because PySocks doesn't support a large number of HTTP proxies. Instead, we recommend using `requests` and its `proxies` keyword argument, which works for HTTP proxies and all types of SOCKS proxies. `requests` uses urllib3's ProxyManager for HTTP proxies, and PySocks only for SOCKS proxies, which should lead to much more reliability.

If not using `requests`, for HTTP proxies I'd recommend using urllib2's native ProxyHandler, and for SOCKS proxies I'd recommend using PySocks' `socket.socksocket` or `socks.setdefaultproxy`.

**Comment by SeaHOH** [comment]
Hi! @Anorov . Thanks for your commentary! 

Have you checked Anorov/PySocks#127 ? We use only one proxy at a time, the type may be one of HTTP or SOCKS, and the target is urllib2, I judge that using `sockshandler.py` is more suitable for this situation. Can you merge the PR?

**Comment by Anorov** [comment]
The issue is that PySocks doesn't really properly support HTTP proxies. In particular, non-CONNECT-based HTTP proxies aren't supported and won't work. This might confuse users with certain HTTP proxies who won't understand why their proxy won't work with this library. I'm not sure how common such proxies are, but we generally recommend using requests because:

* It supports all HTTP proxies and SOCKS proxies
* It's much more widely used and tested for proxying HTTP traffic (compared to urllib2 with sockshandler)
* The API is much simpler to use

I've merged your PR, but I'd just like to make it clear that we don't recommend using urllib2 with PySocks and no longer support urllib2 or sockshandler.

**Comment by SeaHOH** [comment]
Thank you! @Anorov .
urllib2's native ProxyHandler also use HTTP CONNECT tunnelling, I think that non-CONNECT-based HTTP forward proxies are not so common, commonly it used for reverse proxies. requests is really good, but  we don't ues it. Thanks again for your suggest!

**Pull request merged**